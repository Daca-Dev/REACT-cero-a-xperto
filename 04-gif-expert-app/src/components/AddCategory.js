import React, { useState } from 'react';
import PropTypes from 'prop-types'


const AddCategory = ({ setCategories }) => {


    // se deben de usar useStates para poder cambiar y actualizar los valores de variables en el rende de react
    const [inputValue, setInputValue] = useState('');

    const handleInput = (event) => {

        setInputValue( event.target.value );

    }

    const handleSubmit = (event) => {
        /** manejar el evento de submit (presionar enter) en el formulario */
        // evita que el formulario recargue la página, esto ya no se debería aher así
        event.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            // uso de la función pasada en los proptypes por medio de un callback
            setCategories( c => [ inputValue, ...c,] );
            setInputValue('');
        }
        
    }

    return (
        <form onSubmit={ handleSubmit }>
            <input
                type="text"
                value={ inputValue }
                onChange={ handleInput }
                placeholder='Ingrese un texto de búsqueda...'
            />
        </form>
    )
}

// configuracion de los proptypes
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}


export default AddCategory
