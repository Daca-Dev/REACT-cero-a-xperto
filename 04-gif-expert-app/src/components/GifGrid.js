import React from 'react'
import PropTypes from 'prop-types'
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';


export const GifGrid = ({ category }) => {

    
    const { data: images, loading } = useFetchGifs( category );

    return (
        <div className="animate__animated animate__flipInX">
            <h3>{ category }</h3>
            <div className="cards-grid">
                { loading && <p>...Loading</p>}
                {
                    images.map( img => 
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    )
                }
            </div>
        </div>
    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}