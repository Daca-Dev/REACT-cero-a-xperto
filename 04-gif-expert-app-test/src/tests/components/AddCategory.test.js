import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

import { AddCategory } from "../../components/AddCategory";

describe("Prueba del componente <AppCategory />", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("Debe de mostrarse correctamente el componente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("El valor del input debe de cambiar", () => {
    const input = wrapper.find("input");
    const value = "Hola mundo";

    // cambio del valor de simulate y le enviamos un objeto con los datos que usamos
    // en la función que es event.target.value
    input.simulate("change", { target: { value } });

    expect(wrapper.find("p").text().trim()).toBe(value);
  });

  test("No debe de postear la información onSubmit", () => {
    wrapper.find("form").simulate("submit", { preventDefault() {} });
    expect(setCategories).not.toHaveBeenCalled();
  });

  test("Debe de llamar setCategories y limpiar la caja de texto", () => {
    // cambio del valor del input
    const value = "Halo";
    wrapper.find("input").simulate("change", { target: { value } });

    // submit the form
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    // verify that the setCategories has been called
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes( 1 );
    expect(setCategories).toHaveBeenCalledWith( expect.any(Function) );

    // verify that the input value is empty
    expect(wrapper.find("input").prop("value")).toBe('');
  });
});
