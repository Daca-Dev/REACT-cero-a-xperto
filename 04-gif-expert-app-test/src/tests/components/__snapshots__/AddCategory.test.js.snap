// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente <AppCategory /> Debe de mostrarse correctamente el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;

exports[`Pruebas en <AddCategory /> debe de mostrarse correctamente  1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
